File: src/core/ops/group.ts
import type { Registry } from "../../registry/types.ts";
import { MenvError } from "../errors.ts";
import type { OpResult } from "./util.ts";
import { cloneRegistry, newPlan, requireGroup, requireSlug } from "./util.ts";

export function planGroupAdd(registry: Registry, input: { key: string; title: string }): OpResult {
  requireSlug("group", input.key);
  if (registry.groups[input.key] !== undefined) {
    throw new MenvError("VALIDATION", `group "${input.key}" already exists`);
  }
  const next = cloneRegistry(registry);
  next.groups[input.key] = { title: input.title };
  const plan = newPlan();
  plan.registry.push({ action: "set", path: `groups.${input.key}`, summary: `add group "${input.key}"` });
  return { next, plan };
}

export function planGroupUpdate(registry: Registry, input: { key: string; title: string }): OpResult {
  requireGroup(registry, input.key);
  const next = cloneRegistry(registry);
  next.groups[input.key] = { title: input.title };
  const plan = newPlan();
  plan.registry.push({
    action: "set",
    path: `groups.${input.key}.title`,
    summary: `retitle group "${input.key}"`,
  });
  return { next, plan };
}

// Forced outcome: members lose their groupKey (spec's removal table).
export function planGroupRemove(registry: Registry, input: { key: string }): OpResult {
  requireGroup(registry, input.key);
  const next = cloneRegistry(registry);
  const plan = newPlan();
  delete next.groups[input.key];
  plan.registry.push({ action: "remove", path: `groups.${input.key}`, summary: `remove group "${input.key}"` });
  for (const [varName, def] of Object.entries(next.variables)) {
    if (def.groupKey !== input.key) continue;
    def.groupKey = undefined;
    plan.registry.push({
      action: "remove",
      path: `variables.${varName}.groupKey`,
      summary: `clear group on "${varName}"`,
    });
    plan.blockers.push({ code: "GROUP_IN_USE", message: `variable "${varName}" is in group "${input.key}"` });
  }
  return { next, plan };
}

File: go/internal/core/ops/group.go
package ops

import (
	"fmt"
	"sort"

	"github.com/nikrabaev/menv/internal/core"
	"github.com/nikrabaev/menv/internal/registry"
)

// PlanGroupAdd registers a new group.
func PlanGroupAdd(r registry.Registry, input struct{ Key, Title string }) (OpResult, error) {
	if err := RequireSlug("group", input.Key); err != nil {
		return OpResult{}, err
	}
	if _, exists := r.Groups[input.Key]; exists {
		return OpResult{}, &core.MenvError{
			Code:    core.ErrValidation,
			Message: fmt.Sprintf("group %q already exists", input.Key),
		}
	}
	next := CloneRegistry(r)
	next.Groups[input.Key] = registry.GroupDef{Title: input.Title}
	plan := NewPlan()
	plan.Registry = append(plan.Registry, core.RegistryOp{
		Action:  "set",
		Path:    fmt.Sprintf("groups.%s", input.Key),
		Summary: fmt.Sprintf("add group %q", input.Key),
	})
	return OpResult{Next: next, Plan: plan}, nil
}

// PlanGroupUpdate changes a group's title.
func PlanGroupUpdate(r registry.Registry, input struct{ Key, Title string }) (OpResult, error) {
	if _, err := RequireGroup(r, input.Key); err != nil {
		return OpResult{}, err
	}
	next := CloneRegistry(r)
	next.Groups[input.Key] = registry.GroupDef{Title: input.Title}
	plan := NewPlan()
	plan.Registry = append(plan.Registry, core.RegistryOp{
		Action:  "set",
		Path:    fmt.Sprintf("groups.%s.title", input.Key),
		Summary: fmt.Sprintf("retitle group %q", input.Key),
	})
	return OpResult{Next: next, Plan: plan}, nil
}

// PlanGroupRemove removes a group and clears groupKey on all member variables.
func PlanGroupRemove(r registry.Registry, input struct{ Key string }) (OpResult, error) {
	if _, err := RequireGroup(r, input.Key); err != nil {
		return OpResult{}, err
	}
	next := CloneRegistry(r)
	plan := NewPlan()

	delete(next.Groups, input.Key)
	plan.Registry = append(plan.Registry, core.RegistryOp{
		Action:  "remove",
		Path:    fmt.Sprintf("groups.%s", input.Key),
		Summary: fmt.Sprintf("remove group %q", input.Key),
	})

	// Cascade: clear groupKey on all member variables.
	varNames := make([]string, 0, len(next.Variables))
	for n := range next.Variables {
		varNames = append(varNames, n)
	}
	sort.Strings(varNames)
	for _, varName := range varNames {
		def := next.Variables[varName]
		if def.GroupKey != input.Key {
			continue
		}
		def.GroupKey = ""
		next.Variables[varName] = def
		plan.Registry = append(plan.Registry, core.RegistryOp{
			Action:  "remove",
			Path:    fmt.Sprintf("variables.%s.groupKey", varName),
			Summary: fmt.Sprintf("clear group on %q", varName),
		})
		plan.Blockers = append(plan.Blockers, core.PlanIssue{
			Code:    "GROUP_IN_USE",
			Message: fmt.Sprintf("variable %q is in group %q", varName, input.Key),
		})
	}
	return OpResult{Next: next, Plan: plan}, nil
}
